client/mm/mm_arb_market_maker.go

```diff
@@ -330,442 +178,180 @@ func (a *arbMarketMaker) cancelExpiredCEXTrades() {
+	return steppedRate(adjustedRate, mkt.RateStep), nil
+}
+
+// dexPlacementRate calculates the rate at which an order should be placed on
+// the DEX order book based on the rate of the counter trade on the CEX. The
+// logic is in the dexPlacementRate function, so that it can be separately
+// tested.
+func (a *arbMarketMaker) dexPlacementRate(cexRate uint64, sell bool) (uint64, error) {
+	sellFeesInQuoteUnits, err := a.core.OrderFeesInUnits(true, false, cexRate)
 	if err != nil {
-		log.Errorf("error getting quote CEX balance: %v", err)
-		return
+		return 0, fmt.Errorf("error getting sell fees in quote units: %w", err)
 	}
 
-	processSide := func(sell bool) []*rateLots {
-		var cfgPlacements []*ArbMarketMakingPlacement
-		var existingOrders map[int][]*groupedOrder
-		var remainingDEXBalance, remainingCEXBalance, fundingFees uint64
-		if sell {
-			cfgPlacements = cfg.SellPlacements
-			existingOrders = existingSells
-			remainingDEXBalance = baseDEXBalance.Available
-			remainingCEXBalance = quoteCEXBalance.Available
-			fundingFees = sellFees.funding
-		} else {
-			cfgPlacements = cfg.BuyPlacements
-			existingOrders = existingBuys
-			remainingDEXBalance = quoteDEXBalance.Available
-			remainingCEXBalance = baseCEXBalance.Available
-			fundingFees = buyFees.funding
-		}
-
-		cexReserves := reserves.get(!sell, true)
-		if cexReserves > remainingCEXBalance {
-			log.Debugf("rebalance: not enough CEX balance to cover reserves")
-			return nil
-		}
-		remainingCEXBalance -= cexReserves
+	buyFeesInQuoteUnits, err := a.core.OrderFeesInUnits(false, false, cexRate)
+	if err != nil {
+		return 0, fmt.Errorf("error getting buy fees in quote units: %w", err)
+	}
 
-		dexReserves := reserves.get(sell, false)
-		if dexReserves > remainingDEXBalance {
-			log.Debugf("rebalance: not enough DEX balance to cover reserves")
-			return nil
-		}
-		remainingDEXBalance -= dexReserves
-
-		// Enough balance on the CEX needs to be maintained for counter-trades
-		// for each existing trade on the DEX. Here, we reduce the available
-		// balance on the CEX by the amount required for each order on the
-		// DEX books.
-		for _, ordersForPlacement := range existingOrders {
-			for _, o := range ordersForPlacement {
-				var requiredOnCEX uint64
-				if sell {
-					rate := uint64(float64(o.rate) / (1 + cfg.Profit))
-					requiredOnCEX = calc.BaseToQuote(rate, o.lots*mkt.LotSize)
-				} else {
-					requiredOnCEX = o.lots * mkt.LotSize
-				}
-				if requiredOnCEX <= remainingCEXBalance {
-					remainingCEXBalance -= requiredOnCEX
-				} else {
-					log.Warnf("rebalance: not enough CEX balance to cover existing order. cancelling.")
-					addCancel(o)
-					remainingCEXBalance = 0
-				}
-			}
-		}
-		if remainingCEXBalance == 0 {
-			log.Debug("rebalance: not enough CEX balance to place new orders")
-			return nil
-		}
+	return dexPlacementRate(cexRate, sell, a.cfg.Profit, a.mkt, sellFeesInQuoteUnits+buyFeesInQuoteUnits)
+}
```

Comment:
Don't we just need to add or subtract the `halfSpread`?

Reply:
What I had is wrong. It should only be taking into account the fees for either selling or buying, not both. Using the half-spread would be the same as this if initiating and redeeming cost the same amount. It's probably a negligible difference though.

Reply:
The only value that really matters is the spread between our best buy and best sell. Using `halfSpread` would keep the mid-gap the same as the cex. Treating the sides separately effectively shifts the gap if fees are assymetric.

Reply:
Ah I see your point. My logic was that whenever an arb is done, it should cover the fees that are being incurred at that point. I'll just reuse the `halfSpread` code then.

Reply:
> My logic was that whenever an arb is done, it should cover the fees that are being incurred at that point. 

Actually, that kinda sounds right too. Leave it your way.